### ISSUE TITLE
Bert model with attention hints

### ISSUE THREAD

**Issue by pavel-collab** [pavel-collab/SBT-kaggle-nlp#12]
Мы можем сделать модель внутри которой будут подсказки на уровне механизма attention, какие ключевые слова учитывать при классификации.

архитектура модели
```
import torch
import torch.nn as nn
from transformers import BertModel, PreTrainedModel, BertConfig

class BertWithAttentionHints(PreTrainedModel):
    def __init__(self, config):
        super().__init__(config)
        self.bert = BertModel(config)
        self.attn_weights = nn.Linear(config.hidden_size, 1)
        self.classifier = nn.Linear(config.hidden_size, config.num_labels)
        self.init_weights()

    def forward(self, input_ids, attention_mask=None, hint_mask=None, labels=None):
        outputs = self.bert(input_ids=input_ids, attention_mask=attention_mask, return_dict=True)
        token_embeddings = outputs.last_hidden_state  # [B, T, H]

        # Compute attention weights
        raw_scores = self.attn_weights(token_embeddings).squeeze(-1)  # [B, T]

        # Mask out non-hint tokens
        if hint_mask is not None:
            raw_scores = raw_scores.masked_fill(hint_mask == 0, float('-inf'))

        attention_weights = torch.softmax(raw_scores, dim=1)  # [B, T]
        attended = torch.sum(token_embeddings * attention_weights.unsqueeze(-1), dim=1)  # [B, H]

        logits = self.classifier(attended)

        loss = None
        if labels is not None:
            loss = nn.CrossEntropyLoss()(logits, labels)

        return {'loss': loss, 'logits': logits, 'attention': attention_weights}
```

датасет с hint mask
```
class MathTextWithAttentionHints(Dataset):
    def __init__(self, texts, labels, tokenizer, important_words_dict, label2str):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.important_words_dict = important_words_dict
        self.label2str = label2str

    def __getitem__(self, idx):
        text = self.texts[idx]
        label = self.labels[idx]
        encoding = self.tokenizer(text, padding='max_length', truncation=True, max_length=128, return_tensors='pt')
        input_ids = encoding['input_ids'].squeeze()
        attention_mask = encoding['attention_mask'].squeeze()
        tokens = self.tokenizer.convert_ids_to_tokens(input_ids)

        important = self.important_words_dict[self.label2str[label]]
        hint_mask = torch.tensor([1 if any(w in t.lower() for w in important) else 0 for t in tokens])

        return {
            'input_ids': input_ids,
            'attention_mask': attention_mask,
            'hint_mask': hint_mask,
            'labels': torch.tensor(label)
        }

    def __len__(self):
        return len(self.texts)
```

dataloader
```
class DataCollatorWithHintMask:
    def __call__(self, features):
        return {
            'input_ids': torch.stack([f['input_ids'] for f in features]),
            'attention_mask': torch.stack([f['attention_mask'] for f in features]),
            'hint_mask': torch.stack([f['hint_mask'] for f in features]),
            'labels': torch.tensor([f['labels'] for f in features])
        }
```

пример обучения через Trainer
```
model = BertWithAttentionHints(config)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=DataCollatorWithHintMask(),
    tokenizer=tokenizer,
)
trainer.train()
```

**Comment by SanjanShiv** [comment]
Hey @pavel-collab, I have made a PR on this issue. Can you please review it? Thank you.
